pallets/author-mapping/src/lib.rs

```diff
@@ -283,4 +283,126 @@ pub mod pallet {
 			Self::account_and_deposit_of(author_id).map(|info| info.account)
 		}
 	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			use frame_support::storage::migration::{remove_storage_prefix, storage_key_iter};
+			use sp_std::{convert::TryInto, vec::Vec};
+
+			let pallet_prefix: &[u8] = b"AuthorMapping";
+			let storage_item_prefix: &[u8] = b"MappingWithDeposit";
+
+			// Read all the data into memory.
+			// https://crates.parity.io/frame_support/storage/migration/fn.storage_key_iter.html
+			let stored_data: Vec<_> = storage_key_iter::<
+				T::AuthorId,
+				RegistrationInfo<T::AccountId, BalanceOf<T>>,
+				Twox64Concat,
+			>(pallet_prefix, storage_item_prefix)
+			.collect();
+
+			let migrated_count: Weight = stored_data
+				.len()
+				.try_into()
+				.expect("There are between 0 and 2**64 mappings stored.");
+
+			// Now remove the old storage
+			// https://crates.parity.io/frame_support/storage/migration/fn.remove_storage_prefix.html
+			remove_storage_prefix(pallet_prefix, storage_item_prefix, &[]);
+
+			// Assert that old storage is empty
+			assert!(storage_key_iter::<
+				T::AuthorId,
+				RegistrationInfo<T::AccountId, BalanceOf<T>>,
+				Twox64Concat,
+			>(pallet_prefix, storage_item_prefix)
+			.next()
+			.is_none());
+
+			// Write the mappings back to storage with the new secure hasher
+			for (author_id, account_id) in stored_data {
+				MappingWithDeposit::<T>::insert(author_id, account_id);
+			}
+
+			// Return the weight used. For each migrated mapping there is a red to get it into
+			// memory, a write to clear the old stored value, and a write to re-store it.
+			let db_weights = T::DbWeight::get();
+			migrated_count.saturating_mul(2 * db_weights.write + db_weights.read)
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			use frame_support::storage::migration::storage_key_iter;
+			use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+			let pallet_prefix: &[u8] = b"AuthorMapping";
+			let storage_item_prefix: &[u8] = b"MappingWithDeposit";
+
+			// We want to test that:
+			// There are no entries in the new storage beforehand
+			// The same number of mappings exist before and after
+			// As long as there are some mappings stored, one representative key maps to the
+			// same value after the migration.
+			// There are no entries in the old storage afterward
+
+			// Assert new storage is empty
+			//TODO will this actually work? I wonder if the entries from the old map
+			// but the same prefix will confuse it. If so, we could move this assert into the
+			// migration itself to be done after the old storage is cleared.
```

Comment:
```suggestion
			// Because the pallet and item prefixes are the same, the old storage is still at this key. However,
			// the values can't be decoded so the assertion passes.
```